lib/class-wp-rest-url-details-controller.php

```diff
@@ -0,0 +1,260 @@
+		}
+
+		// Transient per URL.
+		$cache_key = $this->build_cache_key_for_url( $url );
+
+		// Attempt to retrieve cached response.
+		$data = $this->get_cache( $cache_key );
+
+		if ( ! empty( $data ) ) {
+			return json_decode( $data, true );
+		}
+
+		$response_body = $this->get_remote_url( $url );
+
+		// Exit if we don't have a valid body or it's empty.
+		if ( is_wp_error( $response_body ) || empty( $response_body ) ) {
+			return $response_body;
+		}
+
+		/**
+		 * Filters the data returned for a given URL.
+		 *
+		 * Can be used to modify the data retrieved for a given URL.
+		 * For example, you may wish to retrieve additional information
+		 * from the response body at a given URL over and above the `<title>`.
+		 *
+		 * @param array $data The default data retrieved for the given URL.
+		 * @param array $response_body The response body (HTML) for the given URL.
+		 * @param string $url The attempted URL.
+		 */
+		$data = apply_filters(
+			'rest_url_details_url_data',
+			array(
+				'title' => $this->get_title( $response_body ),
+			),
+			$response_body,
+			$url
+		);
+
+		// Cache the valid response.
+		$this->set_cache( $cache_key, $data );
+
+		return rest_ensure_response( $data );
+	}
+
+	/**
+	 * Checks whether a given request has permission to read remote urls.
+	 *
+	 * @return WP_Error|bool True if the request has access, WP_Error object otherwise.
+	 */
+	public function permissions_check() {
+		if ( current_user_can( 'edit_posts' ) ) {
+			return true;
+		}
+
+		foreach ( get_post_types( array( 'show_in_rest' => true ), 'objects' ) as $post_type ) {
+			if ( current_user_can( $post_type->cap->edit_posts ) ) {
+				return true;
+			}
+		}
+
+		return new WP_Error(
+			'rest_user_cannot_view',
+			__( 'Sorry, you are not allowed to process remote urls.', 'gutenberg' ),
+			array( 'status' => rest_authorization_required_code() )
+		);
+	}
+
+	/**
+	 * Retrieves the document title from a remote URL.
+	 *
+	 * @param string $url The website url whose HTML we want to access.
+	 * @return string|WP_Error The URL's document title on success, WP_Error on failure.
+	 */
+	private function get_remote_url( $url ) {
+
+		$args = array(
+			'limit_response_size' => 150 * KB_IN_BYTES,
+		);
+
+		/**
+		 * Filters the HTTP request args for URL data retrieval.
+		 *
+		 * Can be used to adjust response size limit and other WP_Http::request args.
+		 *
+		 * @param array $args Arguments used for the HTTP request
+		 * @param string $url The attempted URL.
+		 */
+		$args = apply_filters( 'rest_url_details_http_request_args', $args, $url );
+
+		$response = wp_safe_remote_get(
+			$url,
+			$args
+		);
+
+		if ( WP_Http::OK !== wp_remote_retrieve_response_code( $response ) ) {
+			// Not saving the error response to cache since the error might be temporary.
+			return new WP_Error( 'no_response', get_status_header_desc( 404 ), array( 'status' => WP_Http::NOT_FOUND ) );
```

Comment:
`get_status_header_desc` is used for the HTTP status code description and isn't translated. We should use a specific error message of our own here.

Reply:
Updated